core/src/main/java/ai/timefold/solver/core/impl/bavet/common/RecordAndReplayPropagator.java

```diff
@@ -87,33 +84,51 @@ public void retract(Object object) {
     @Override
     public void propagateRetracts() {
         if (!retractQueue.isEmpty() || !insertQueue.isEmpty()) {
+            var precomputeBuildHelper = precomputeBuildHelperSupplier.get();
+            var internalNodeNetwork = precomputeBuildHelper.getNodeNetwork();
+            var recordingTupleLifecycle = precomputeBuildHelper.getRecordingTupleLifecycle();
+
             updateQueue.removeAll(retractQueue);
             updateQueue.removeAll(insertQueue);
+
+            var allObjectSet = CollectionUtils
+                    .newIdentityHashSet(insertQueue.size() + objectToOutputTuplesMap.size());
+
             // Do not remove queued retracts from inserts; if a fact property
             // change, there will be both a retract and insert for that fact
+
+            allObjectSet.addAll(objectToOutputTuplesMap.keySet());
+            allObjectSet.removeAll(retractQueue);
```

Comment:
I am not so sure this is as efficient as we think.

In order to be able to do a bulk `addAll`, we have to do `removeAll` with some other collection. IMO a much simpler "add if not contains" in a loop would result in overall less `add` operations, with the amount of `contains` remaining the same (since `removeAll` does something similar anyway).